Mathlib/Algebra/GroupWithZero/Power.lean

```diff
@@ -15,6 +15,25 @@ In this file we define integer power functions for groups with an adjoined zero
 This generalises the integer power function on a division ring.
 -/
 
+section MonoidWithZero
+
+variable {M₀ : Type _} [MonoidWithZero M₀] {x y : M₀} {k : ℕ}
+
+/-- If `x` is not a (left) zero divisor then neither is `x^k` for any `k`. -/
+lemma eq_zero_of_forall_mul_eq_zero_of_mul_pow_eq_zero_left
+    (h : ∀ y, y * x = 0 → y = 0) (h' : y * (x^k) = 0) : y = 0 := by
+  induction' k with l ih; simpa using h'
+  rw [pow_succ', ← mul_assoc] at h'
+  exact ih (h _ h')
+
+/-- If `x` is not a (right) zero divisor then neither is `x^k` for any `k`. -/
```

Comment:
Let me try again: I think that more generally, the product of any two non right zero divisors is a non right zero divisor.  That is, the set of non right zero divisors is a semigroup.

Unless I am still confused, the proof is just associativity and application twice of the non-zero divisor hypothesis.

Reply:
This is a great suggestion. I have dropped these lemmas an instead proposed introducing `nonZeroDivisorsLeft` and `nonZeroDivisorsRight` so that I can get this result via `Submonoid.pow_mem` which is essentially what you're suggesting. Thanks!